llvm/lib/IR/RepoTicket.cpp

```diff
@@ -118,76 +127,122 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
 
 // Update the GO's hash value by adding the hash of its dependents.
 template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
+static auto updateDigestUseDependenciesAndContributions(
+    const GlobalObject *GO, GOStateMap &Visited, GOInfoMap &GOIMap,
+    MemoizedHashes &HashTable, Function AccumulateGODigest)
+    -> std::tuple<bool, size_t, DigestType> {
   if (GO->isDeclaration())
-    return false;
+    return std::make_tuple(false, std::numeric_limits<size_t>::max(),
+                           NullDigest);
+
+  const auto NumVisited = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << NumVisited << ")\n");
+
+  // If the hash has been memorized, we can return the result immediately.
+  const auto TablePos = HashTable.find(GO);
+  if (TablePos != HashTable.end()) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(true, NumVisited, TablePos->second);
+  }
+
+  MD5 GOHash = MD5();
 
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, NumVisited);
   if (!Inserted) {
+    assert(NumVisited > StateIt->second);
     // If GO is visited, use the letter 'R' as the marker and use its state as
     // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    GOHash.update(static_cast<uint8_t>(Tags::Backref));
+    GOHash.update(NumVisited - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    MD5::MD5Result Digest;
+    GOHash.final(Digest);
+    return std::make_tuple(true, StateIt->second, Digest);
   }
 
   bool Changed = false;
-  GOHash.update('T');
+  GOHash.update(static_cast<uint8_t>(Tags::GO));
   auto State = AccumulateGODigest(GO, GOHash, GOIMap);
   Changed = Changed || State.Changed;
-  // Update the GO's hash using its contributions and dependencies.
-  auto &GOContributions = State.Contributions;
-  auto &GODependencies = State.Dependencies;
-  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
-           make_range(GOContributions.begin(), GOContributions.end()),
-           make_range(GODependencies.begin(), GODependencies.end()))) {
+  // Update the GO's hash using its contributions. Contributions don't form the
+  // loop since 1) only the global variable has the contributions and 2) the
+  // contributions are the functions.
```

Comment:
Sorry I don't understand this comment. 

> Update the GO's hash using its contributions.

Replace with "Add the GO's contributions to its hash"?

> Contributions don't form the loop since 1) only the global variable has the contributions

Replace with "Contributions can't form loops since 1) only global variables have contributions"? Could we assert that this is true?

> and 2) the contributions are the functions.

I can't guess what you mean here!

Reply:
>
> Replace with "Add the GO's contributions to its hash"?
> 
Fixed, thanks.
>
> Replace with "Contributions can't form loops since 1) only global variables have contributions"? 
> 
Fixed, thanks.
>
>Could we assert that this is true?
>
Two assertions are added here.

> 
> I can't guess what you mean here!
>
The GO’s contributions are collected from the `ContributedToGVs`, which is a vector of global variables to which the function has contributed. Therefore, only functions have been put into the contributions of global variables.